examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+            hash: None,
+        },
+    }))?;
+
+    Ok(state)
+}
+
+/// Initialzes the implementation of the w_ccd contract with the proxy and the
+/// state addresses.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "initialize",
+    parameter = "InitializeImplementationParams",
+    mutable
+)]
+fn contract_initialize<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation<S>, StateApiType = S>,
+) -> ContractResult<()> {
+    ensure!(!host.state().initialized, concordium_cis2::Cis2Error::Custom(AlreadyInitialized));
+    // Set proxy and storage addresses.
+    let params: InitializeImplementationParams = ctx.parameter_cursor().get()?;
+    host.state_mut().proxy_address = params.proxy_address;
+    host.state_mut().state_address = params.state_address;
+    // Set the w_ccd_implementation contract as initialized.
+    host.state_mut().initialized = true;
+    Ok(())
+}
+
+/// Wrap an amount of CCD into wCCD tokens and transfer the tokens if the sender
+/// is not the receiver.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "wrap",
+    parameter = "WrapParams",
+    enable_logger,
+    mutable,
+    payable
+)]
+fn contract_wrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation<S>, StateApiType = S>,
+    amount: Amount,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    let slot_time = ctx.metadata().slot_time();
+
+    match host.state().state_address {
+        None => bail!(concordium_cis2::Cis2Error::Custom(OnlyState)),
+
+        Some(state_address) => {
+            let unpause_time = host
+                .invoke_contract_raw(
+                    &state_address,
+                    Parameter(&[]),
+                    EntrypointName::new_unchecked("get_unpause_time"),
+                    Amount::zero(),
+                )?
+                .1;
+
+            let unpause_time = if let Some(mut unpause_time) = unpause_time {
+                unpause_time.get()?
+            } else {
+                return Err(concordium_cis2::Cis2Error::Custom(StateInvokeError));
+            };
+
+            ensure!(slot_time >= unpause_time, concordium_cis2::Cis2Error::Custom(ContractPaused));
+
+            let params: WrapParams = ctx.parameter_cursor().get()?;
+            // Get the sender who invoked this contract function.
+            let sender = ctx.sender();
+
+            let receive_address = params.to.address();
+
+            let (state, state_builder) = host.state_and_builder();
+
+            // Update the state.
+            state.mint(&TOKEN_ID_WCCD, amount.micro_ccd.into(), &receive_address, state_builder)?;
+
+            // Log the newly minted tokens.
+            logger.log(&Cis2Event::Mint(MintEvent {
+                token_id: TOKEN_ID_WCCD,
+                amount:   ContractTokenAmount::from(amount.micro_ccd),
+                owner:    sender,
+            }))?;
```

Comment:
I think I'm missing something, but why do we first log that the owner is the sender, and then log the transfer.

Why don't we just log that upon minting they are owned by the `receive_address`.

Reply:
Well, that would also be a solution.

If we chose to log a mint event directly to the receiving address, the logs would not show the information of the sender address who actually paid for this.